src/web/js/trove/chart-lib.js

```diff
@@ -174,6 +214,116 @@
 
   //////////////////////////////////////////////////////////////////////////////
 
+  /**
+   * Adds multiple columns with the given properties and values after data
+   * columns
+   * 
+   * For example, if given:
+   * colProperties:
+   *   {type: 'string', role: 'style'}
+   * colValues:
+   *   [
+   *     [['red', 'black'], ['white', 'blue'], ['green', 'purple']],
+   *     []
+   *   ]
+   * addNSpecialColumns will add 2 style columns after the first data column
+   * and no columns after the second data column.
+   * 
+   * The number of columns added after a particular data column do not have to
+   * agree. It is possible to add one special value on one row and two
+   * special values on another row.
+   * 
+   * https://jsfiddle.net/eyanje/u83kaf92/
+   * 
+   * @param {DataTable} table a table to expand
+   * @param {object} colProperties an object specifying column properties
+   * @param {Array<Array<*>>>} colValues rows of groups of values to insert
+   */
+  function addNSpecialColumns(table, colProperties, colValues) {
+    let dataColNums = [];
+    let nDataCols;
+    let groupWidths;
+    for (let i = 1; i < table.getNumberOfColumns(); i++) {
+      const role = table.getColumnRole(i);
+      if (role === '' || role === 'data') {
+        dataColNums.push(i);
+      }
+    }
+    nDataCols = dataColNums.length;
+    // Check column count
+    // Should never run -- Pyret checks all column counts properly
+    // This should be somewhat caught in the try-catch around setup(restarter),
+    // unless it's been moved
+    colValues.forEach((row, rowN) => {
+      if (row.length !== nDataCols) {
+        throw new Error(`Incorrect column count in row ${rowN}.`
+          + ` Expected ${nDataCols}, given ${row.length}.`);
+      }
+    });
+    // Tally columns needed for each group
+    groupWidths = dataColNums.map(() => 0);
+    colValues.forEach(row => {
+      row.forEach((group, groupN) => {
+        groupWidths[groupN] = Math.max(group.length, groupWidths[groupN]);
+      });
+    });
+    // Add columns in reverse order
+    for (let groupIndex = nDataCols - 1; groupIndex >= 0; groupIndex--) {
+      for (let i = 0; i < groupWidths[groupIndex]; i++) {
+        table.insertColumn(dataColNums[groupIndex] + 1, colProperties);
+      }
+    }
+    // Adjust dataColNums to match expanded table
+    let sum = 0;
+    dataColNums.forEach((dataColNum, i) => {
+        dataColNums[i] += sum;
+      sum += groupWidths[i];
+    });
+    // Add columns in reverse order to avoid extra calculations
+    colValues.forEach((row, rowN) => {
+        row.forEach((group, groupN) => {
+          group.forEach((val, i) => {
+          table.setValue(rowN, dataColNums[groupN] + i + 1, val);
+        });
+      })
+    });
+  }
+
+  /**
```

Comment:
For example, I think we wrote this function, but the block above came from somewhere else – that's useful to know when debugging later.

Reply:
Yeah, but since the example is specific to the function, I think it actually makes sense to keep the jsfiddle comment above the function and maybe make a copy reference to it at the top if you want all the jsfiddles to be referenced at the top.